test/common/upstream/load_balancer_impl_test.cc

```diff
@@ -916,6 +916,23 @@ TEST_P(RoundRobinLoadBalancerTest, WeightedSeed) {
   EXPECT_EQ(hostSet().healthy_hosts_[1], lb_->chooseHost(nullptr));
 }
 
+// Validate that low weight hosts still have the corresponding opportunity to be picked first.
+TEST_P(RoundRobinLoadBalancerTest, WeightedBiasAvoiding) {
+  hostSet().healthy_hosts_ = {makeTestHost(info_, "tcp://127.0.0.1:80", 98),
+                              makeTestHost(info_, "tcp://127.0.0.1:81", 2)};
+  hostSet().hosts_ = hostSet().healthy_hosts_;
+  // Initiate current_time to 0.49999, and make two hosts' initial deadlines to be both 0.5.
+  EXPECT_CALL(random_, random()).WillRepeatedly(Return(49999));
```

Comment:
In addition to this test where you return the same random value but get differing hosts returned between first and subsequent host selections, it wouldn't be too difficult to do something like:
```
for (int i = 0; i < 10000; ++i) {
   EXPECT_CALL(random_, random()).WillRepeatedly(Return(i));
   init(false);
  // track the host returned by chooseHost
}
```

and verify that your first selections don't contain a bias. The selection probabilities should come out pretty close to what the weights specify.

Reply:
OK, I made a test case, and finally the probability is 2.04% for the host that weights 2, which is pretty close to 2%.

Upload modified test soon.